Jaspr Lint Quick Fix Not working

### Discussed in https://github.com/schultek/jaspr/discussions/405

<sup>Originally posted by **TJMusiitwa** March 23, 2025</sup>
Hey @schultek 
I have the latest `jaspr_lints` package installed and part of my dev dependencies. And with that I am getting several `sort_children_properties_last` lint warnings being raised. The docs mention that there is a quick fix available. However, when I hover over the bulb icon or just do a Save action in VS Code, this _fix_ is not applied.

Best course of action to address this, as going in and manually doing the edits in over **10** different files in my project, with **178** lint warnings reported seems a bit too much.

**Jaspr Doctor Output**

> [✓] Jaspr CLI (Version 0.18.0)
>   • Dart Version 3.7.2 (stable) (Tue Mar 11 04:27:50 2025 -0700) on "macos_arm64" at /Users/jonathan/development/flutter/bin/cache/dart-sdk/bin/dart
>   • Running on macos Version 14.1.2 (Build 23B92) - Locale en-GB
>   • Analytics: Enabled
> 
> [✓] Current Project
>   • Dependencies on core packages:
>     • jaspr: ^0.18.0
>     • jaspr_builder: ^0.18.0 (dev)
>   • Rendering mode: static
>   • Uses jaspr compilers: false
>   • Uses flutter embedding: false

  
  **Pubspec.yaml**
```
environment:
  sdk: '>=3.7.2 <4.0.0'

dependencies:
  jaspr: ^0.18.0

dev_dependencies:
  build_runner: ^2.4.15
  build_web_compilers: ^4.1.3
  flutter_lints: ^5.0.0
  jaspr_builder: ^0.18.0
  jaspr_lints: ^0.3.1
  # jaspr_tailwind: ^0.3.1

jaspr:
  mode: static
```

---

You can try jaspr analyze --fix

---

Hey so running this command leads to a massive error in the terminal

```
[ERROR] Plugin SortChildrenPropertiesLastFix threw while analyzing /Users/jonathan/Desktop/conference/lib/components/banner.dart:
[ERROR] ConflictingEditException: {"offset":702,"length":0,"replacement":", [\n            div([\n              p(\n                [\n                  text(\n                    'Registration has closed for this year\\'s event. Stay tuned for updates on next year\\'s event.',\n                  ),\n                ],\n                classes:\n                    'me-2 inline-block text-sm font-bold text-gray-800 dark:text-neutral-200',\n              ),\n            ], classes: 'space-x-4 items-center justify-center'),\n          ]"} conflicts with {"offset":181,"length":543,"replacement":""}
[ERROR] #0      addEditForSource (package:analyzer_plugin/src/protocol/protocol_internal.dart:77:7)
[ERROR] #1      SourceFileEdit.add (package:analyzer_plugin/protocol/protocol_common.dart:4193:7)
[ERROR] #2      FileEditBuilderImpl._addEdit (package:analyzer_plugin/src/utilities/change_builder/change_builder_core.dart:623:14)
[ERROR] #3      FileEditBuilderImpl._addEditBuilder (package:analyzer_plugin/src/utilities/change_builder/change_builder_core.dart:638:5)
[ERROR] #4      FileEditBuilderImpl.addInsertion (package:analyzer_plugin/src/utilities/change_builder/change_builder_core.dart:541:7)
[ERROR] #5      DartFileEditBuilderImpl.addInsertion (package:analyzer_plugin/src/utilities/change_builder/change_builder_dart.dart:1990:13)
[ERROR] #6      SortChildrenPropertiesLastFix.run.<anonymous closure> (package:jaspr_lints/src/lints/sort_children_properties_last_lint.dart:58:17)
[ERROR] #7      ChangeBuilderImpl.addDartFileEdit (package:analyzer_plugin/src/utilities/change_builder/change_builder_core.dart:169:26)
[ERROR] #8      ChangeBuilderImpl.addDartFileEdit.<anonymous closure> (package:custom_lint_core/src/change_reporter.dart:241:35)
[ERROR] #9      new Future.<anonymous closure> (dart:async/future.dart:260:40)
[ERROR] #10     _rootRun (dart:async/zone.dart:1517:47)
[ERROR] #11     _CustomZone.run (dart:async/zone.dart:1422:19)
[ERROR] #12     _CustomZone.runGuarded (dart:async/zone.dart:1321:7)
[ERROR] #13     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1362:23)
[ERROR] #14     _rootRun (dart:async/zone.dart:1525:13)
[ERROR] #15     _CustomZone.run (dart:async/zone.dart:1422:19)
[ERROR] #16     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1345:23)
[ERROR] #17     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
[ERROR] #18     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:410:19)
[ERROR] #19     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:441:5)
[ERROR] #20     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:194:12)
```
And this is my banner file

```dart
import 'package:jaspr/jaspr.dart';

class Banner extends StatelessComponent {
  @override
  Iterable<Component> build(BuildContext context) sync* {
    yield div([
      a(
        [
          div([
            div([
              p(
                [
                  text(
                    'Registration has closed for this year\'s event. Stay tuned for updates on next year\'s event.',
                  ),
                ],
                classes:
                    'me-2 inline-block text-sm font-bold text-gray-800 dark:text-neutral-200',
              ),
            ], classes: 'space-x-4 items-center justify-center'),
          ], classes: 'max-w-[85rem] px-4 sm:px-6 lg:px-8 mx-auto'),
        ],
        classes:
            'group block bg-white/60 backdrop-blur-sm hover:bg-gray-200 focus:outline-none focus:bg-gray-200 p-4 rounded-lg text-center transition duration-300',
        target: Target.blank,
        href: 'https://tally.so/r/w598xZ',
      ),
    ]);
  }
}
```

The process does not continue as I assume it starts in files alphabetical order so the banner file is the first of those file and as such it just remains in that state.
Unsure how to proceed from here.

--- a/packages/jaspr_lints/CHANGELOG.md
+++ b/packages/jaspr_lints/CHANGELOG.md
@@ -1,3 +1,7 @@
+## Unreleased patch
+
+- Fixed conflicting analysis error for "Sort children last" quick fix.
+
 ## 0.5.0
 
 - Extend `prefer_html_methods` lint to also suggest using `text()` and `fragment()` methods, and offer fixes for them.

--- a/packages/jaspr_lints/lib/src/lints/sort_children_properties_last_lint.dart
+++ b/packages/jaspr_lints/lib/src/lints/sort_children_properties_last_lint.dart
@@ -53,17 +53,22 @@ class SortChildrenPropertiesLastFix extends DartFix {
     if (analysisError.data case (InvocationExpression node, ListLiteral childrenArg)) {
       reporter.createChangeBuilder(message: 'Sort children last', priority: 2).addDartFileEdit((builder) {
         var content = resolver.source.contents.data;
-        var lines = content.substring(childrenArg.offset, childrenArg.end);
-
-        builder.addInsertion(node.argumentList.rightParenthesis.offset, (edit) {
-          if (node.argumentList.rightParenthesis.previous?.lexeme != ',') {
-            edit.write(', ');
-          }
-          edit.write(lines);
-        });
 
         var nextArg = node.argumentList.arguments[node.argumentList.arguments.indexOf(childrenArg) + 1];
-        builder.addDeletion(SourceRange(childrenArg.offset, nextArg.offset - childrenArg.offset));
+
+        var start = childrenArg.end;
+        var argStart = nextArg.offset;
+        var end = node.argumentList.arguments.last.end;
+
+        var divider = content.substring(start, argStart);
+        var args = content.substring(argStart, end);
+
+        builder.addInsertion(childrenArg.offset, (edit) {
+          edit.write(args);
+          edit.write(divider);
+        });
+
+        builder.addDeletion(SourceRange(start, end - start));
       });
     }
   }

